### ISSUE TITLE
difference in typecheck with 5.8 (broken type inference)

### ISSUE THREAD

**Issue by dartess** [microsoft/TypeScript#63775]
> Same errors, locations, and messages as TS5.8

I got 2 "new" errors on my huge project. I consider it an incredible luck, but I think I should report on the remaining two. here is the first one:

(This is the minimum example I could reproduce with the original error. As soon as I simplify it more I either get another error or they disappear)

```ts
type Message = { options?: Record<string, unknown> };

type Shared<T> = { handler: (message: T) => void };
type Static<T> = Shared<T> & { message: T };
type Dynamic<T> = Shared<T> & { message: () => T };

declare function subscribe<TMessage extends Message>(
  payload: (Dynamic<TMessage> | Static<TMessage>) & {},
): void;

subscribe({
  message: () => ({
    options: { market: '' },
  }),
  handler: (message) => message.options.market,
});

export {};

```

```
error TS18048: 'message.options' is possibly 'undefined'.

15   handler: (message) => message.options.market,
```

TS 5.8.3 is fine with this: [playground](https://www.typescriptlang.org/play/?#code/C4TwDgpgBAshDO8CGBzaBeKBvKB7MwAlrgHbwD8AXFAEoQDGuATgCYA88wThJKANFACuJANYlcAdxIA+KAF8A3ACgloSFADKACyRMI7ACqzMOHSRYAbCE2oAKALYJkaagYCUUdLIBuuQi3llNWgNYCQiejYjT00dPUNZADJsKEdEVAhXQNVwaAAREBIke0JI6MxtXX0opJS050yoWw8vKANspRYGCyqoADNheiJSKHhBACN4em5xiCi4dLQoCAAPYAhzeFgnDOlbJSgoMCQQC1wkFjsCopKyhYbZAB9NMIj5nbRpD2SsOT4lNzUXz+ZRKMaTaaEWa2LAHVIfRrNTyyGFww74YZkag4ey6EQQYDUADkRPk-0Ocjc5KgZks1js9QyLVkjLQADoMcQyGzcUx8cA2cBcABVMCQJgAYSQ8AgzX+lNBqzAzGA2EUSiAA)

The easiest way to "fix" this error is to remove the intersection `& {}` (but in my code there is a type with some fields and I can't)

**Comment by birgersp** [comment]
I just ran into a similar issue when using parameter properties and I think (hopefully) it is relevant for this issue. In package A, I have this class definition:

```
export class SomeClass {
	// "timestamp" is a parameter property
	constructor(readonly timestamp = new Date()) {}
}
```

And in package B, I'll get this (false positive?) error:

```
const o = new SomeClass()
console.log(o.timestamp.getTime()) // <-- error TS18048: 'o.timestamp' is possibly 'undefined'.
```

However if I change the class definition to:

```
export class SomeClass {
	// property is no longer declared by a parameter property
	readonly timestamp: Date
	constructor(timestamp = new Date()) {
		this.timestamp = timestamp
	}
}
```

... the error goes away.

For what it's worth I'm on version 7.0.0-dev.20250704.1

It seems that the error only pops up when the class definition is in a different package 🤔

**Comment by birgersp** [comment]
Fiddling around with it some more, I see that this code:

```
export class SomeClass {
	constructor(readonly timestamp = new Date()) {}
}
```

Is compiled to two different things between tsc (5.8.3) and tsgo (7.0.0-dev.20250704.1).

tsc result ([fiddle](https://www.typescriptlang.org/play/?ts=5.8.3#code/KYDwDg9gTgLgBAYwDYEMDOa4GUIFtgDCqGcA3gFACQCEAdmjFAK4IzQAUUwKAJnUgE84MAJb4GKXGDgBeOLWAB3OABEUMYOwCUWsgF9yeoA)):

```
export declare class SomeClass {
    readonly timestamp: Date;
    constructor(timestamp?: Date);
}
```

tsgo result:

```
export declare class SomeClass {
    readonly timestamp: Date | undefined;
    constructor(timestamp?: Date | undefined);
}
```

So, perhaps not the same issue(?)

**Comment by jakebailey** [comment]
Not the same issue, no. You're correct to file separately in microsoft/typescript-go#1478.

**Comment by ahejlsberg** [comment]
As described above, this is a type ordering issue and TS 6 reports the same error with `-stableTypeOrdering`.

**Comment by dartess** [comment]
The issue was opened as a potential update issue, and since this is expected behavior, I think it should be closed.
